File: dashboard-flutter/lib/services/server_api.dart
import '../core/api_client.dart';

/// أغلفة نقاط الـ API الخادمية للوحة (كلّها محميّة بـ Bearer Firebase ID token).
/// تطابق `internetArchive.js` / `noorLibrary.js` / `hindawiLibrary.js` /
/// `crawl4ai.js` / `supervisors.js` / `admin.js`.
class ServerApi {
  ServerApi._();
  static final ServerApi instance = ServerApi._();
  final _api = ApiClient.instance;

  // ─── محرّك عام (Internet Archive / Noor / Hindawi) ───────

  /// مسار الأساس لكل محرّك.
  static String engineBase(EngineKind kind) => switch (kind) {
    EngineKind.internetArchive => '/api/admin/internet-archive/engine',
    EngineKind.noor => '/api/admin/noor-library/engine',
    EngineKind.hindawi => '/api/admin/hindawi-library/engine',
  };

  Future<dynamic> engineStatus(EngineKind kind, {int logLimit = 30}) {
    final base = engineBase(kind);
    final path = kind == EngineKind.internetArchive
        ? '$base/status'
        : '$base/status?limit=$logLimit';
    return _api.authedJson(path);
  }

  Future<dynamic> engineStart(EngineKind kind) =>
      _api.authedJson('${engineBase(kind)}/start', method: 'POST', body: {});

  Future<dynamic> engineStop(EngineKind kind) =>
      _api.authedJson('${engineBase(kind)}/stop', method: 'POST', body: {});

  Future<dynamic> engineTick(EngineKind kind) =>
      _api.authedJson('${engineBase(kind)}/tick', method: 'POST', body: {});

  /// إعادة الضبط. IA يستعمل {type}، Noor/Hindawi يستعملان {mode}.
  Future<dynamic> engineReset(EngineKind kind, {String type = 'cursor'}) {
    final base = engineBase(kind);
    if (kind == EngineKind.internetArchive) {
      return _api.authedJson(
        '$base/reset',
        method: 'POST',
        body: {'type': type},
      );
    }
    return _api.authedJson('$base/reset', method: 'POST', body: {'mode': type});
  }

  // ─── Internet Archive (خاص) ─────────────────────────────

  Future<dynamic> iaBootstrap() => _api.authedJson(
    '/api/admin/internet-archive/engine/bootstrap',
    method: 'POST',
    body: {},
  );

  Future<dynamic> iaDiagnose() => _api.authedJson(
    '/api/admin/internet-archive/engine/diagnose',
    method: 'POST',
    body: {},
  );

  Future<dynamic> iaSectionsTree() =>
      _api.authedJson('/api/admin/internet-archive/sections');

  Future<dynamic> iaSearch(Map<String, dynamic> args) => _api.authedJson(
    '/api/admin/internet-archive/search',
    method: 'POST',
    body: args,
  );

  Future<dynamic> iaPreview(Map<String, dynamic> args) => _api.authedJson(
    '/api/admin/internet-archive/preview',
    method: 'POST',
    body: args,
  );

  Future<dynamic> iaImport(Map<String, dynamic> args) => _api.authedJson(
    '/api/admin/internet-archive/import',
    method: 'POST',
    body: args,
  );

  // ─── Noor / Hindawi diagnose ────────────────────────────

  Future<dynamic> librariesDiagnose(String source) =>
      _api.authedJson('/api/admin/libraries/diagnose?source=$source');

  Future<dynamic> noorSectionsTree() =>
      _api.authedJson('/api/admin/noor-library/sections');

  Future<dynamic> noorPreview(String url) => _api.authedJson(
    '/api/admin/noor-library/preview',
    method: 'POST',
    body: {'url': url},
  );

  Future<dynamic> noorImport(Map<String, dynamic> payload) => _api.authedJson(
    '/api/admin/noor-library/import',
    method: 'POST',
    body: payload,
  );

  Future<dynamic> noorJobs({int limit = 30}) =>
      _api.authedJson('/api/admin/noor-library/jobs?limit=$limit');

  // ─── crawl4ai ───────────────────────────────────────────

  Future<dynamic> crawl4aiStatus() =>
      _api.authedJson('/api/admin/crawl4ai/status');

  Future<dynamic> crawl4aiControl(String action) => _api.authedJson(
    '/api/admin/crawl4ai/control',
    method: 'POST',
    body: {'action': action},
  );

  Future<dynamic> crawl4aiEnqueue(String url) => _api.authedJson(
    '/api/admin/crawl4ai/crawl',
    method: 'POST',
    body: {'url': url},
  );

  Future<dynamic> crawl4aiJobs({int limit = 50}) =>
      _api.authedJson('/api/admin/crawl4ai/jobs?limit=$limit');

  // ─── المشرفون ───────────────────────────────────────────

  Future<dynamic> listSupervisors() =>
      _api.authedJson('/api/admin/supervisors');

  Future<dynamic> setSupervisorBlocked(
    String uid,
    bool isBlocked, {
    String mode = 'temporary',
  }) => _api.authedJson(
    '/api/admin/supervisors/${Uri.encodeComponent(uid)}',
    method: 'PATCH',
    body: {'isBlocked': isBlocked, 'mode': mode},
  );

  Future<dynamic> removeSupervisor(String uid) => _api.authedJson(
    '/api/admin/supervisors/${Uri.encodeComponent(uid)}',
    method: 'DELETE',
  );

  // ─── إحصاءات / تقارير / تدقيق ───────────────────────────

  Future<dynamic> aggregatesOverview() =>
      _api.authedJson('/api/admin/aggregates/overview');

  Future<dynamic> aggregatesBySource() =>
      _api.authedJson('/api/admin/aggregates/by-source');

  Future<dynamic> reports({Map<String, String>? query}) =>
      _api.authedJson('/api/admin/reports', query: query);

  /// إجراء على بلاغ: {action:'dismiss'|'delete', reportId, contentId, contentType?}
  Future<dynamic> reportAction(Map<String, dynamic> body) =>
      _api.authedJson('/api/admin/reports', method: 'POST', body: body);

  /// قنوات ومنشورات «مجتمع نبراس» التي تحتاج إلى إشراف.
  Future<dynamic> communityModeration() =>
      _api.authedJson('/api/admin/community');

  /// تعليق قناة أو فك تعليقها، وحذف منشور مخالف وفق صلاحية المستخدم.
  Future<dynamic> communityAction(Map<String, dynamic> body) =>
      _api.authedJson('/api/admin/community', method: 'POST', body: body);

  Future<dynamic> contentAudit({Map<String, String>? query}) =>
      _api.authedJson('/api/admin/content-audit', query: query);

  /// حذف عنصر مُعلَّم من تدقيق المحتوى.
  Future<dynamic> contentAuditDelete(String contentId, String contentType) =>
      _api.authedJson(
        '/api/admin/content-audit',
        method: 'POST',
        body: {
          'action': 'delete',
          'contentId': contentId,
          'contentType': contentType,
        },
      );

  // ─── إشعارات FCM ────────────────────────────────────────

  Future<dynamic> notify(Map<String, dynamic> body) =>
      _api.authedJson('/api/notify', method: 'POST', body: body);
}

enum EngineKind { internetArchive, noor, hindawi }

File: nebras-dashboard/app/src/main/kotlin/com/nebras/dashboard/services/ServerApi.kt
package com.nebras.dashboard.services

import android.net.Uri
import com.nebras.dashboard.core.ApiClient

/** المحرّكات الثلاثة — نظير `enum EngineKind` في Dart. */
enum class EngineKind { internetArchive, noor, hindawi }

/**
 * أغلفة نقاط الـ API الخادمية للوحة (كلّها محميّة بـ Bearer Firebase ID token).
 * تطابق `internetArchive.js` / `noorLibrary.js` / `hindawiLibrary.js` /
 * `crawl4ai.js` / `supervisors.js` / `admin.js`.
 *
 * كلّ دالّة `suspend` وتُرجع الحمولة المفكوكة (Map/List/قيمة أوّليّة أو null)
 * تماماً كما كانت `Future<dynamic>` في Dart.
 */
object ServerApi {

    /** نظير `ServerApi.instance` في Dart — نفس المفرد. */
    val instance: ServerApi get() = this

    // ─── محرّك عام (Internet Archive / Noor / Hindawi) ───────

    /** مسار الأساس لكل محرّك. */
    fun engineBase(kind: EngineKind): String = when (kind) {
        EngineKind.internetArchive -> "/api/admin/internet-archive/engine"
        EngineKind.noor -> "/api/admin/noor-library/engine"
        EngineKind.hindawi -> "/api/admin/hindawi-library/engine"
    }

    suspend fun engineStatus(kind: EngineKind, logLimit: Int = 30): Any? {
        val base = engineBase(kind)
        val path = if (kind == EngineKind.internetArchive) {
            "$base/status"
        } else {
            "$base/status?limit=$logLimit"
        }
        return ApiClient.authedJson(path)
    }

    suspend fun engineStart(kind: EngineKind): Any? =
        ApiClient.authedJson("${engineBase(kind)}/start", method = "POST", body = emptyMap<String, Any?>())

    suspend fun engineStop(kind: EngineKind): Any? =
        ApiClient.authedJson("${engineBase(kind)}/stop", method = "POST", body = emptyMap<String, Any?>())

    suspend fun engineTick(kind: EngineKind): Any? =
        ApiClient.authedJson("${engineBase(kind)}/tick", method = "POST", body = emptyMap<String, Any?>())

    /** إعادة الضبط. IA يستعمل {type}، Noor/Hindawi يستعملان {mode}. */
    suspend fun engineReset(kind: EngineKind, type: String = "cursor"): Any? {
        val base = engineBase(kind)
        if (kind == EngineKind.internetArchive) {
            return ApiClient.authedJson(
                "$base/reset",
                method = "POST",
                body = mapOf("type" to type),
            )
        }
        return ApiClient.authedJson("$base/reset", method = "POST", body = mapOf("mode" to type))
    }

    // ─── Internet Archive (خاص) ─────────────────────────────

    suspend fun iaBootstrap(): Any? = ApiClient.authedJson(
        "/api/admin/internet-archive/engine/bootstrap",
        method = "POST",
        body = emptyMap<String, Any?>(),
    )

    suspend fun iaDiagnose(): Any? = ApiClient.authedJson(
        "/api/admin/internet-archive/engine/diagnose",
        method = "POST",
        body = emptyMap<String, Any?>(),
    )

    suspend fun iaSectionsTree(): Any? =
        ApiClient.authedJson("/api/admin/internet-archive/sections")

    suspend fun iaSearch(args: Map<String, Any?>): Any? = ApiClient.authedJson(
        "/api/admin/internet-archive/search",
        method = "POST",
        body = args,
    )

    suspend fun iaPreview(args: Map<String, Any?>): Any? = ApiClient.authedJson(
        "/api/admin/internet-archive/preview",
        method = "POST",
        body = args,
    )

    suspend fun iaImport(args: Map<String, Any?>): Any? = ApiClient.authedJson(
        "/api/admin/internet-archive/import",
        method = "POST",
        body = args,
    )

    // ─── Noor / Hindawi diagnose ────────────────────────────

    suspend fun librariesDiagnose(source: String): Any? =
        ApiClient.authedJson("/api/admin/libraries/diagnose?source=$source")

    suspend fun noorSectionsTree(): Any? =
        ApiClient.authedJson("/api/admin/noor-library/sections")

    suspend fun noorPreview(url: String): Any? = ApiClient.authedJson(
        "/api/admin/noor-library/preview",
        method = "POST",
        body = mapOf("url" to url),
    )

    suspend fun noorImport(payload: Map<String, Any?>): Any? = ApiClient.authedJson(
        "/api/admin/noor-library/import",
        method = "POST",
        body = payload,
    )

    suspend fun noorJobs(limit: Int = 30): Any? =
        ApiClient.authedJson("/api/admin/noor-library/jobs?limit=$limit")

    // ─── crawl4ai ───────────────────────────────────────────

    suspend fun crawl4aiStatus(): Any? =
        ApiClient.authedJson("/api/admin/crawl4ai/status")

    suspend fun crawl4aiControl(action: String): Any? = ApiClient.authedJson(
        "/api/admin/crawl4ai/control",
        method = "POST",
        body = mapOf("action" to action),
    )

    suspend fun crawl4aiEnqueue(url: String): Any? = ApiClient.authedJson(
        "/api/admin/crawl4ai/crawl",
        method = "POST",
        body = mapOf("url" to url),
    )

    suspend fun crawl4aiJobs(limit: Int = 50): Any? =
        ApiClient.authedJson("/api/admin/crawl4ai/jobs?limit=$limit")

    // ─── المشرفون ───────────────────────────────────────────

    suspend fun listSupervisors(): Any? =
        ApiClient.authedJson("/api/admin/supervisors")

    suspend fun setSupervisorBlocked(
        uid: String,
        isBlocked: Boolean,
        mode: String = "temporary",
    ): Any? = ApiClient.authedJson(
        "/api/admin/supervisors/${Uri.encode(uid)}",
        method = "PATCH",
        body = mapOf("isBlocked" to isBlocked, "mode" to mode),
    )

    suspend fun removeSupervisor(uid: String): Any? = ApiClient.authedJson(
        "/api/admin/supervisors/${Uri.encode(uid)}",
        method = "DELETE",
    )

    // ─── إحصاءات / تقارير / تدقيق ───────────────────────────

    suspend fun aggregatesOverview(): Any? =
        ApiClient.authedJson("/api/admin/aggregates/overview")

    suspend fun aggregatesBySource(): Any? =
        ApiClient.authedJson("/api/admin/aggregates/by-source")

    suspend fun reports(query: Map<String, String>? = null): Any? =
        ApiClient.authedJson("/api/admin/reports", query = query)

    /** إجراء على بلاغ: {action:'dismiss'|'delete', reportId, contentId, contentType?} */
    suspend fun reportAction(body: Map<String, Any?>): Any? =
        ApiClient.authedJson("/api/admin/reports", method = "POST", body = body)

    /** قنوات ومنشورات «مجتمع نبراس» التي تحتاج إلى إشراف. */
    suspend fun communityModeration(): Any? =
        ApiClient.authedJson("/api/admin/community")

    /** تعليق قناة أو فك تعليقها، وحذف منشور مخالف وفق صلاحية المستخدم. */
    suspend fun communityAction(body: Map<String, Any?>): Any? =
        ApiClient.authedJson("/api/admin/community", method = "POST", body = body)

    suspend fun contentAudit(query: Map<String, String>? = null): Any? =
        ApiClient.authedJson("/api/admin/content-audit", query = query)

    /** حذف عنصر مُعلَّم من تدقيق المحتوى. */
    suspend fun contentAuditDelete(contentId: String, contentType: String): Any? =
        ApiClient.authedJson(
            "/api/admin/content-audit",
            method = "POST",
            body = mapOf(
                "action" to "delete",
                "contentId" to contentId,
                "contentType" to contentType,
            ),
        )

    // ─── إشعارات FCM ────────────────────────────────────────

    suspend fun notify(body: Map<String, Any?>): Any? =
        ApiClient.authedJson("/api/notify", method = "POST", body = body)
}
